pkg/graveler/ref/branch_iterator_test.go

```diff
@@ -94,3 +96,191 @@ func TestBranchIterator(t *testing.T) {
+func TestBranchByCommitIterator(t *testing.T) {
+	r, kvStore := testRefManagerWithKV(t)
+	// TODO niro: Need commits PR (remove 'main' from list when done)
+	branches := []graveler.BranchID{"a", "aa", "b", "c", "e", "d", "main"}
+	ctx := context.Background()
+	repo := &graveler.RepositoryRecord{
+		RepositoryID: "repo1",
+		Repository: &graveler.Repository{
+			StorageNamespace: "s3://foo",
+			CreationDate:     time.Now(),
+			DefaultBranchID:  "main",
+		},
+	}
+	testutil.Must(t, r.CreateRepository(ctx, repo.RepositoryID, *repo.Repository, ""))
+
+	// prepare data
+	for i, b := range branches {
+		testutil.Must(t, r.SetBranch(ctx, repo.RepositoryID, b, graveler.Branch{CommitID: graveler.CommitID(branches[len(branches)-i-1])}))
+	}
+
+	t.Run("listing all branches", func(t *testing.T) {
+		iter, err := ref.NewBranchByCommitIterator(ctx, &kvStore, repo.RepositoryID.String())
+		require.NoError(t, err)
+		ids := make([]graveler.CommitID, 0)
+		for iter.Next() {
+			b := iter.Value()
+			ids = append(ids, b.CommitID)
+		}
+		fmt.Println(ids)
+		if iter.Err() != nil {
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+		iter.Close()
+
+		if diffs := deep.Equal(ids, []graveler.CommitID{"a", "aa", "b", "c", "d", "e", "main"}); diffs != nil {
+			t.Fatalf("got wrong list of IDs: %v", diffs)
+		}
+	})
+
+	t.Run("listing branches using prefix", func(t *testing.T) {
+		iter, err := ref.NewBranchByCommitIterator(ctx, &kvStore, repo.RepositoryID.String())
+		require.NoError(t, err)
+		iter.SeekGE("b")
+		ids := make([]graveler.CommitID, 0)
+		for iter.Next() {
+			b := iter.Value()
+			ids = append(ids, b.CommitID)
+		}
+		if iter.Err() != nil {
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+		iter.Close()
+
+		if diffs := deep.Equal(ids, []graveler.CommitID{"b", "c", "d", "e", "main"}); diffs != nil {
+			t.Fatalf("got wrong list of branch IDs: %v", diffs)
+		}
+	})
+
+	t.Run("listing branches SeekGE", func(t *testing.T) {
+		iter, err := ref.NewBranchByCommitIterator(ctx, &kvStore, repo.RepositoryID.String())
+		require.NoError(t, err)
+		iter.SeekGE("b")
+		ids := make([]graveler.CommitID, 0)
+		for iter.Next() {
+			b := iter.Value()
+			ids = append(ids, b.CommitID)
+		}
+		if iter.Err() != nil {
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+
+		if diffs := deep.Equal(ids, []graveler.CommitID{"b", "c", "d", "e", "main"}); diffs != nil {
+			t.Fatalf("got wrong list of branch IDs: %v", diffs)
+		}
+
+		// now let's seek
+		iter.SeekGE("aa")
+		ids = make([]graveler.CommitID, 0)
+		for iter.Next() {
+			b := iter.Value()
+			ids = append(ids, b.CommitID)
+		}
+		if iter.Err() != nil {
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+		iter.Close()
+
+		if diffs := deep.Equal(ids, []graveler.CommitID{"aa", "b", "c", "d", "e", "main"}); diffs != nil {
+			t.Fatalf("got wrong list of branch IDs")
+		}
+
+		require.False(t, iter.Next())
+		require.Nil(t, iter.Value())
+		require.ErrorIs(t, iter.Err(), ref.ErrIteratorClosed)
+	})
+}
```

Comment:
Might be interesting to benchmark `BranchByCommitIterator` against the old `BranchIterator` with `orderByCommitID = true`, just to get a sense of how painful the sort is.
If it is not that significant we can peacefully go on, but if it is too painful, we might want to consider different approach.
Either ways, I think this insight can be useful

Reply:
It would - though we will need to be very careful testing, the benchmark should include the creation + iteration flows since the sorting happens during the creation process.
Added comment in: https://github.com/treeverse/lakeFS/issues/3571